src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,253 @@
+        }
+
+        /// <summary>
+        /// Subscribes to the underlying <see cref="IObservable{T}"/> with the
+        /// transformation specified by this instance.
+        /// <br/><br/>
+        /// Disconnection requests via the returned <see cref="IDisposable"/> interface
+        /// are passed to the underlying <see cref="IObservable{T}"/> and also used
+        /// to signal pending asynchronous tasks that cancellation has been requested
+        /// and also used to prevent further event notifications.
+        /// </summary>
+        public IDisposable Subscribe(IObserver<TOut> observer)
+        {
+            IDisposable? disposable = null;
+            var newObserver = new Observer(observer, _transformNext, _transformError, () => disposable?.Dispose());
+            disposable = _observable.Subscribe(newObserver);
+            return newObserver;
+        }
+
+        private class Observer : IObserver<TIn>, IDisposable
+        {
+            private CancellationTokenSource? _cancellationTokenSource = new();
+            private readonly CancellationToken _token;
+            //create a queue so that events will be sent in order
+            private readonly Queue<QueueData> _queue = new();
+            private readonly IObserver<TOut> _observer;
+            private readonly Func<TIn, CancellationToken, Task<TOut>> _transformNext;
+            private readonly Func<Exception, CancellationToken, Task<Exception>> _transformError;
+            private Action? _disposeAction;
+
+            public Observer(IObserver<TOut> observer, Func<TIn, CancellationToken, Task<TOut>> transformNext, Func<Exception, CancellationToken, Task<Exception>> transformError, Action disposeAction)
+            {
+                _token = _cancellationTokenSource.Token;
+                _observer = observer;
+                _disposeAction = disposeAction;
+                // ensure that the transform cannot directly throw an exception without it being wrapped in a Task<Exception>
+                _transformError = async (exception, token) => await transformError(exception, token).ConfigureAwait(false);
+                // ensure that the transform cannot directly throw an exception without it being wrapped in a Task<TOut>
+                _transformNext = async (data, token) => await transformNext(data, token).ConfigureAwait(false);
+            }
+
+            public void OnNext(TIn value) => QueueEvent(QueueType.Data, _transformNext(value, _token), null);
+
+            public void OnError(Exception error) => QueueEvent(QueueType.Error, null, _transformError(error, _token));
+
+            public void OnCompleted() => QueueEvent(QueueType.Completion, null, null);
+
+            /// <summary>
+            /// Queues the specified event and if necessary starts watching for an event to complete.
+            /// </summary>
+            private void QueueEvent(QueueType queueType, Task<TOut>? task, Task<Exception>? error)
```

Comment:
🤔 I understand what you are doing here and it seems to me that it mey be simplified
> As a side note, the re-sequencing within SelectCatchAsync was accomplished without a dedicated thread or synchronization primitives except shortly-held locks, and code is executed synchronously when possible.

I can not yet say exactly in what form I would prefer to see the processing of the queue. I will come back later.

Reply:
I'm ready here. You already simplified this code a bit. I'm talking about `attach` flag. It is a valve for the start of the queue processing. The problem is that `ProcessAllEventsInQueueAsync` may be started more than 1 time (2, 10, even 100). For this, it is enough that the queue is initially empty, and the `ProcessAllEventsInQueueAsync` method has performing a long work (for example, inside `ProcessDataAsync`). So the control flows into `ProcessAllEventsInQueueAsync` again since `attach` is `true`. Now we have two concurrent tasks inside `ProcessAllEventsInQueueAsync`. And they _may_ handle the same element (both call `TryPeek`) more that 1 time. In theory this can happen again and again (of course probability ~0). I suggest to remove `attach` local flag at all and switch to field like `_queueRunning`. Then examine this flag via `Interlocked` APIs. `ProcessAllEventsInQueueAsync` stays the very same except adding `Interlocked.CompareExchange` at the very beginning as a guard of concurrent calls into `ProcessAllEventsInQueueAsync`.

Reply:
I do not understand.  It is absolutely impossible for `ProcessAllEventsInQueueAsync` to be running concurrently with another `ProcessAllEventsInQueueAsync`.  Either it attaches to the head of an empty queue, and so starts executing, or the queue has something in it and it does not start executing.   Within the method, if it empties the queue (within a lock) it is guaranteed to immediately exit the method.  It is quite impossible for multiple executions to be simultaneously running.

This is very very important to execute correctly, so let me know if I missed something.

I do spot a bug: unhandled exceptions may break the `ProcessAllEventsInQueueAsync` loop and prevent continued execution of events in the loop.

Reply:
It is important when you look at the code to realize that `ProcessAllEventsInQueueAsync` does NOT remove an item at the top of the method.  It peeks it, leaving the queue with a Count > 0, until processing has finished.  Then at the bottom of the method it removes it from the queue and (determined within the lock) either quits because Count == 0 or continues because Count > 0

Reply:
You are right. I checked again. Pretty good implementation.